src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            //   do { ... } while (YieldProcessor());
+            Helpers.Spin(1);
+            return true;
+        }
+
+        private ManualResetEvent GetOrCreateReaderEvent()
+        {
+            ManualResetEvent currentEvent = _readerEvent;
+            if (currentEvent != null)
+            {
+                return currentEvent;
+            }
+
+            currentEvent = new ManualResetEvent(false);
+            ManualResetEvent previousEvent = Interlocked.CompareExchange(ref _readerEvent, currentEvent, null);
+            if (previousEvent == null)
+            {
+                return currentEvent;
+            }
+
+            currentEvent.Dispose();
+            return previousEvent;
+        }
+
+        private AutoResetEvent GetOrCreateWriterEvent()
+        {
+            AutoResetEvent currentEvent = _writerEvent;
+            if (currentEvent != null)
+            {
+                return currentEvent;
+            }
+
+            currentEvent = new AutoResetEvent(false);
+            AutoResetEvent previousEvent = Interlocked.CompareExchange(ref _writerEvent, currentEvent, null);
+            if (previousEvent == null)
+            {
+                return currentEvent;
+            }
+
+            currentEvent.Dispose();
+            return previousEvent;
+        }
+
+        private void ReleaseEvents()
+        {
+            Debug.Assert((_state & LockStates.CachingEvents) == LockStates.CachingEvents);
+
+            // Save events
+            AutoResetEvent writerEvent = _writerEvent;
+            _writerEvent = null;
+            ManualResetEvent readerEvent = _readerEvent;
+            _readerEvent = null;
+
+            // Allow readers and writers to continue
+            Interlocked.Add(ref _state, -LockStates.CachingEvents);
+
+            // Cache events
+            // TODO: (old) Disposing events for now. What is needed is an event cache to which the events are released.
+            if (writerEvent != null)
+            {
+                writerEvent.Dispose();
+            }
+            if (readerEvent != null)
+            {
+                readerEvent.Dispose();
+            }
+
+#if RWLOCK_STATISTICS
+            Interlocked.Increment(ref _eventsReleasedCount);
+#endif
+        }
+
+        private static ArgumentOutOfRangeException GetInvalidTimeoutException(string parameterName)
+        {
+            return new ArgumentOutOfRangeException(parameterName, SR.ArgumentOutOfRange_TimeoutMilliseconds);
+        }
+
+        private static int ToTimeoutMilliseconds(TimeSpan timeout)
+        {
+            var timeoutMilliseconds = (long)timeout.TotalMilliseconds;
+            if (timeoutMilliseconds < -1 || timeoutMilliseconds > int.MaxValue)
+            {
+                throw GetInvalidTimeoutException(nameof(timeout));
+            }
+            return (int)timeoutMilliseconds;
+        }
+
+        private class ReaderWriterLockApplicationException : ApplicationException
```

Comment:
What exception type is thrown in the desktop implementation, a plain ApplicationException?  If so, consider marking this exception type [Serialiable] and adding a deserialization ctor that just calls to the base.  I'm thinking of a case where someone was catching the exception previously and serializing it; with this new derived exception type, that would fail.